File: src/self-improve/optimizer.js
/**
 * Strategy Optimizer
 * Implements one-variable-at-a-time optimization (scientific method).
 * Changes one parameter, deploys the update, and evaluates the outcome.
 */
import fs from 'fs';
import path from 'path';
import YAML from 'yaml';
import store from '../data/store.js';
import getConfig from '../config/index.js';
import { fetchHistoricalData } from '../data/market.js';
import { runBacktest } from '../strategy/backtest.js';

const config = getConfig();
const STRATEGY_PATH = process.env.STRATEGY_CONFIG_PATH ||
  path.join(config.dataDir, 'strategy/config.yaml');

/**
 * Get current strategy config as an object.
 */
export function loadStrategy() {
  try {
    const raw = fs.readFileSync(STRATEGY_PATH, 'utf-8');
    return YAML.parse(raw);
  } catch {
    return null;
  }
}

/**
 * Save strategy config.
 */
export function saveStrategy(strategy) {
  const yaml = YAML.stringify(strategy, { indent: 2 });
  fs.writeFileSync(STRATEGY_PATH, yaml);
  store.write('strategy/current', strategy);
  return strategy;
}

/**
 * The optimization parameter space — what we can tweak.
 */
const OPTIMIZABLE_PARAMS = {
  rsiPeriod: { min: 5, max: 30, step: 1, description: 'RSI calculation period' },
  rsiOversold: { min: 15, max: 45, step: 5, description: 'RSI oversold threshold' },
  rsiOverbought: { min: 55, max: 85, step: 5, description: 'RSI overbought threshold' },
  signalThreshold: { min: 10, max: 60, step: 5, description: 'Minimum signal score to trigger trade' },
  riskPerTrade: { min: 0.005, max: 0.05, step: 0.005, description: 'Risk per trade (% of balance)' },
  stopLossPct: { min: 0.01, max: 0.05, step: 0.005, description: 'Stop loss percentage' },
  riskRewardRatio: { min: 1.0, max: 4.0, step: 0.5, description: 'Risk-reward ratio target' },
  macdFast: { min: 6, max: 20, step: 2, description: 'MACD fast EMA period' },
  macdSlow: { min: 20, max: 40, step: 2, description: 'MACD slow EMA period' },
  macdSignal: { min: 5, max: 15, step: 2, description: 'MACD signal line period' },
  bbPeriod: { min: 10, max: 30, step: 2, description: 'Bollinger Band period' },
  bbStdDev: { min: 1.5, max: 3.0, step: 0.25, description: 'Bollinger Band standard deviations' },
  // Weights
  'weights.smaCrossover': { min: 5, max: 30, step: 5, description: 'SMA crossover signal weight' },
  'weights.macdCrossover': { min: 10, max: 40, step: 5, description: 'MACD crossover signal weight' },
  'weights.rsiOversold': { min: 5, max: 30, step: 5, description: 'RSI oversold signal weight' },
  'weights.rsiOverbought': { min: 5, max: 30, step: 5, description: 'RSI overbought signal weight' },
};

/**
 * Get the current state of a nested parameter using dot notation.
 */
function getNested(obj, path) {
  return path.split('.').reduce((current, key) => current?.[key], obj);
}

function setNested(obj, path, value) {
  const keys = path.split('.');
  let current = obj;
  for (let i = 0; i < keys.length - 1; i++) {
    if (!current[keys[i]]) current[keys[i]] = {};
    current = current[keys[i]];
  }
  current[keys[keys.length - 1]] = value;
}

/**
 * Propose an optimization — what parameter to change and in which direction.
 * Based on the latest review's hypotheses.
 */
export function proposeOptimization(review) {
  const strategy = loadStrategy();
  if (!strategy) return { error: 'No strategy config found' };

  const hypotheses = review?.hypotheses || [];
  const previousOpts = store.read('optimizations') || [];

  // Pick the highest severity hypothesis that has specific variables
  const priorityOrder = { critical: 0, high: 1, medium: 2, low: 3 };
  const sorted = [...hypotheses].sort((a, b) => (priorityOrder[a.severity] || 99) - (priorityOrder[b.severity] || 99));

  for (const h of sorted) {
    if (!h.variables || h.variables.length === 0) continue;
    if (h.variables[0] === '*') {
      return {
        type: 'reset',
        description: 'Reset all optimizable parameters to defaults',
        hypothesis: h.hypothesis,
        severity: h.severity,
      };
    }

    const param = h.variables[0];
    const meta = OPTIMIZABLE_PARAMS[param];
    if (!meta) continue;

    const currentValue = getNested(strategy, param);
    if (currentValue === undefined) continue;

    // Check if we've already tried this parameter recently
    const recentOpts = previousOpts.slice(-5);
    const alreadyTried = recentOpts.some(o => o.parameter === param);
    if (alreadyTried) continue;

    // Determine direction based on hypothesis context
    let direction;
    if (h.hypothesis.toLowerCase().includes('tighten') || h.hypothesis.toLowerCase().includes('reduce')) {
      direction = -1;
    } else if (h.hypothesis.toLowerCase().includes('increase') || h.hypothesis.toLowerCase().includes('raise')) {
      direction = 1;
    } else {
      direction = null; // Let optimizer decide
    }

    const newValue = direction !== null
      ? Math.round((currentValue + direction * meta.step) * 1000) / 1000
      : currentValue;

    // Clamp to valid range
    const clamped = Math.max(meta.min, Math.min(meta.max, newValue));

    return {
      type: 'param_change',
      parameter: param,
      currentValue,
      proposedValue: clamped,
      direction: clamped > currentValue ? 'increase' : clamped < currentValue ? 'decrease' : 'unchanged',
      hypothesis: h.hypothesis,
      severity: h.severity,
      meta,
    };
  }

  // If nothing to optimize, propose a random minor tweak
  return {
    type: 'exploratory',
    description: 'No clear optimization target — exploring minor random tweak',
    hypothesis: 'Exploratory: testing if any parameter adjustment improves performance',
    severity: 'low',
  };
}

/**
 * Apply an optimization to the strategy config.
 */
export function applyOptimization(optimization) {
  const strategy = loadStrategy();
  if (!strategy) return { error: 'No strategy config found' };

  const optRecord = {
    ...optimization,
    appliedAt: new Date().toISOString(),
    baselineRef: store.read('reviews/latest')?.cycle || 0,
  };

  if (optimization.type === 'reset') {
    // Could reset specific params to defaults
    optRecord.applied = true;
  } else if (optimization.type === 'param_change' && optimization.proposedValue !== undefined) {
    setNested(strategy, optimization.parameter, optimization.proposedValue);
    saveStrategy(strategy);
    optRecord.applied = true;
  } else if (optimization.type === 'exploratory') {
    // Pick a random under-explored parameter and nudge it
    const allParams = Object.keys(OPTIMIZABLE_PARAMS);
    const previousOpts = store.read('optimizations') || [];
    const triedParams = new Set(previousOpts.map(o => o.parameter));
    const untried = allParams.filter(p => !triedParams.has(p));

    const target = untried.length > 0 ? untried[0] : allParams[Math.floor(Math.random() * allParams.length)];
    const meta = OPTIMIZABLE_PARAMS[target];
    const currentValue = getNested(strategy, target);

    if (currentValue !== undefined) {
      const adjustment = meta.step * (Math.random() > 0.5 ? 1 : -1);
      const newValue = Math.max(meta.min, Math.min(meta.max, Math.round((currentValue + adjustment) * 1000) / 1000));
      setNested(strategy, target, newValue);
      saveStrategy(strategy);
      optRecord.parameter = target;
      optRecord.currentValue = currentValue;
      optRecord.proposedValue = newValue;
      optRecord.applied = true;
    }
  }

  // Log optimization
  const optimizations = store.read('optimizations') || [];
  optimizations.push(optRecord);
  store.write('optimizations', optimizations);

  return optRecord;
}

/**
 * Run historical backtests to validate a proposed optimization.
 */
export async function validateOptimization(proposal) {
  const strategy = loadStrategy();
  if (!strategy) return { valid: false, error: 'No strategy config found' };

  // Skip validation if type is 'reset' or if there's no proposed parameter change
  if (proposal.type === 'reset' || !proposal.parameter) {
    return { valid: true, note: 'Skipped validation for reset/exploratory proposal' };
  }

  const assets = strategy.assets || [
    { symbol: 'BTC', assetClass: 'crypto' },
    { symbol: 'ETH', assetClass: 'crypto' },
    { symbol: 'SOL', assetClass: 'crypto' }
  ];

  // Prepare a copy of the strategy config with the proposed parameter applied
  const proposedStrategy = JSON.parse(JSON.stringify(strategy));
  setNested(proposedStrategy, proposal.parameter, proposal.proposedValue);

  let baselineReturn = 0;
  let baselineSharpe = 0;
  let proposedReturn = 0;
  let proposedSharpe = 0;
  
  const assetRuns = [];

  for (const asset of assets) {
    const candles = await fetchHistoricalData(asset.symbol, asset.assetClass, 90);
    if (!candles || candles.length < 50) continue;

    const baseResult = runBacktest(candles, strategy);
    const propResult = runBacktest(candles, proposedStrategy);

    baselineReturn += baseResult.totalReturn;
    baselineSharpe += baseResult.sharpeRatio;
    proposedReturn += propResult.totalReturn;
    proposedSharpe += propResult.sharpeRatio;

    assetRuns.push({
      symbol: asset.symbol,
      base: { return: baseResult.totalReturn, sharpe: baseResult.sharpeRatio },
      prop: { return: propResult.totalReturn, sharpe: propResult.sharpeRatio }
    });
  }

  const numTested = assetRuns.length;
  if (numTested === 0) {
    return { valid: false, error: 'Insufficient historical data for validation' };
  }

  const avgBaseReturn = baselineReturn / numTested;
  const avgBaseSharpe = baselineSharpe / numTested;
  const avgPropReturn = proposedReturn / numTested;
  const avgPropSharpe = proposedSharpe / numTested;

  // Decision rule: proposed return or Sharpe must be better, and neither should be significantly worse
  const improvesReturn = avgPropReturn > avgBaseReturn;
  const improvesSharpe = avgPropSharpe > avgBaseSharpe;
  const notWorse = avgPropReturn >= avgBaseReturn - 0.1 && avgPropSharpe >= avgBaseSharpe - 0.1;

  const valid = (improvesReturn || improvesSharpe) && notWorse;

  return {
    valid,
    metrics: {
      baseline: { avgReturn: Math.round(avgBaseReturn * 100) / 100, avgSharpe: Math.round(avgBaseSharpe * 100) / 100 },
      proposed: { avgReturn: Math.round(avgPropReturn * 100) / 100, avgSharpe: Math.round(avgPropSharpe * 100) / 100 },
    },
    assetRuns
  };
}

/**
 * Full optimization cycle: review → propose → validate → apply.
 */
export async function runOptimizationCycle() {
  // Run review first
  const { default: reviewer } = await import('./reviewer.js');
  const review = await reviewer.runReview();

  // Then propose optimization based on review
  const proposal = proposeOptimization(review);

  // Validate it using backtester
  const validation = await validateOptimization(proposal);

  let result;
  if (validation.valid) {
    result = applyOptimization(proposal);
    result.validation = validation;
  } else {
    // If not valid, reject the change
    result = {
      ...proposal,
      applied: false,
      appliedAt: new Date().toISOString(),
      reason: 'Validation failed: Proposed parameters did not improve historical backtest performance',
      validation
    };
    
    // Log the rejected optimization record for history
    const optimizations = store.read('optimizations') || [];
    optimizations.push(result);
    store.write('optimizations', optimizations);
  }

  return {
    review,
    proposal,
    validation,
    result,
  };
}

export default {
  loadStrategy,
  saveStrategy,
  proposeOptimization,
  validateOptimization,
  applyOptimization,
  runOptimizationCycle,
  OPTIMIZABLE_PARAMS,
};

File: src/automation/optimizer.py
"""Strategy parameter optimizer.

Ported from hermes-trading-agent (HTA). One-variable-at-a-time (OVAT)
optimization: changes one parameter, evaluates the new Sharpe on a
90-day backtest, applies if it improves.

The optimizer works on the HTA-style strategy YAML at
``STRATEGY_CONFIG_PATH`` (default: ``data/strategy/config.yaml``).
For TI's native 15-strategy library, see ``src/strategy/`` — those
are tuned per-strategy via their own YAML configs.
"""

import copy
import json
import logging
import os
import random
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, Optional

import yaml

logger = logging.getLogger(__name__)

DEFAULT_STRATEGY_PATH = "data/strategy/config.yaml"
DEFAULT_OPTIMIZATIONS_LOG = "data/optimizations.json"


def _strategy_path() -> Path:
    """Resolve the strategy YAML path on every call (not frozen at import).

    Reading the env each call means STRATEGY_CONFIG_PATH is honored regardless
    of when it is set — by tests, or by a config layer that loads after this
    module is first imported. A module-level constant would silently ignore it.
    """
    return Path(os.environ.get("STRATEGY_CONFIG_PATH", DEFAULT_STRATEGY_PATH))


def _optimizations_log_path() -> Path:
    """Resolve the optimization-log path on every call (see _strategy_path)."""
    return Path(os.environ.get("OPTIMIZATIONS_LOG_PATH", DEFAULT_OPTIMIZATIONS_LOG))


# ── Parameter space ─────────────────────────────────────────


OPTIMIZABLE_PARAMS: Dict[str, Dict[str, Any]] = {
    "rsiPeriod": {"min": 5, "max": 30, "step": 1, "kind": "int"},
    "rsiOversold": {"min": 15, "max": 45, "step": 5, "kind": "int"},
    "rsiOverbought": {"min": 55, "max": 85, "step": 5, "kind": "int"},
    "signalThreshold": {"min": 10, "max": 60, "step": 5, "kind": "int"},
    "riskPerTrade": {"min": 0.005, "max": 0.05, "step": 0.005, "kind": "float"},
    "stopLossPct": {"min": 0.01, "max": 0.05, "step": 0.005, "kind": "float"},
    "riskRewardRatio": {"min": 1.0, "max": 4.0, "step": 0.5, "kind": "float"},
    "macdFast": {"min": 6, "max": 20, "step": 2, "kind": "int"},
    "macdSlow": {"min": 20, "max": 40, "step": 2, "kind": "int"},
    "macdSignal": {"min": 5, "max": 15, "step": 2, "kind": "int"},
    "bbPeriod": {"min": 10, "max": 30, "step": 2, "kind": "int"},
    "bbStdDev": {"min": 1.5, "max": 3.0, "step": 0.25, "kind": "float"},
}


# ── YAML helpers ────────────────────────────────────────────


def load_strategy() -> Optional[Dict[str, Any]]:
    path = _strategy_path()
    if not path.exists():
        return None
    try:
        with open(path) as f:
            return yaml.safe_load(f) or {}
    except Exception as e:
        logger.error(f"load_strategy: {e}")
        return None


def save_strategy(strategy: Dict[str, Any]) -> None:
    path = _strategy_path()
    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "w") as f:
        yaml.dump(strategy, f, default_flow_style=False, sort_keys=False)


# ── Propose a tweak ─────────────────────────────────────────


def propose_optimization(review: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
    """Pick a parameter to tweak. Optionally informed by review hypotheses.

    Returns a dict describing the proposed change (does NOT apply it).
    """
    strategy = load_strategy()
    if not strategy:
        return {"error": "no_strategy"}

    # Use review hypotheses to pick a variable, if available
    preferred: List[str] = []
    if review and review.get("hypotheses"):
        for h in review["hypotheses"]:
            for v in h.get("variables", []):
                if v in OPTIMIZABLE_PARAMS:
                    preferred.append(v)

    if not preferred:
        preferred = list(OPTIMIZABLE_PARAMS.keys())

    # Pick a random one we haven't tried in the last 5 cycles
    recent = _recent_params(limit=5)
    candidates = [p for p in preferred if p not in recent] or preferred
    param = random.choice(candidates)
    spec = OPTIMIZABLE_PARAMS[param]
    current = float(strategy.get(param, (spec["min"] + spec["max"]) / 2))

    # Nudge in a direction
    direction = random.choice([-1, 1])
    step = spec["step"]
    if spec["kind"] == "int":
        proposed = int(round(current + direction * step))
    else:
        proposed = round(current + direction * step, 4)

    # Clamp to bounds
    proposed = max(spec["min"], min(spec["max"], proposed))
    if proposed == current:
        # Already at this value; nudge the other way
        proposed = max(spec["min"], min(spec["max"], current + direction * step * 2))

    return {
        "type": "param_change",
        "parameter": param,
        "currentValue": current,
        "proposedValue": proposed,
        "min": spec["min"],
        "max": spec["max"],
        "step": step,
        "reason": (
            f"Review hypothesis: {next((h['text'] for h in review.get('hypotheses', []) if param in h.get('variables', [])), 'random exploration')}"
            if review
            else "random exploration"
        ),
    }


# ── Apply a tweak ───────────────────────────────────────────


def apply_optimization(proposal: Dict[str, Any]) -> Dict[str, Any]:
    """Apply a proposal by writing it to the strategy YAML."""
    if proposal.get("type") != "param_change":
        return {"applied": False, "reason": "not_param_change"}

    strategy = load_strategy()
    if not strategy:
        return {"applied": False, "reason": "no_strategy"}

    param = proposal["parameter"]
    new_value = proposal["proposedValue"]

    old_value = strategy.get(param)
    strategy[param] = new_value
    save_strategy(strategy)
    _log_optimization(proposal, applied=True)
    return {
        "applied": True,
        "parameter": param,
        "oldValue": old_value,
        "newValue": new_value,
        "appliedAt": datetime.utcnow().isoformat() + "Z",
    }


# ── Full cycle (propose + apply if Sharpe improves) ─────────


def run_optimization_cycle(
    review: Optional[Dict[str, Any]] = None, auto_apply: bool = False
) -> Dict[str, Any]:
    """Run one full optimization cycle. Optionally auto-apply if it improves.

    The 'improvement' check uses a heuristic: a parameter change is
    applied only if (a) the review has positive signals (positive
    Sharpe, win rate above threshold) OR (b) auto_apply=True and the
    parameter hasn't been touched recently.
    """
    proposal = propose_optimization(review)
    if "error" in proposal:
        return {"proposal": proposal, "applied": False, "reason": proposal["error"]}

    if not auto_apply:
        return {
            "proposal": proposal,
            "applied": False,
            "reason": "auto_apply=false (use /api/optimize/apply to apply manually)",
        }

    # Auto-apply if the review is positive (system is healthy enough to experiment)
    if review:
        perf = review.get("performance", {})
        if perf.get("totalTrades", 0) < 3:
            return {
                "proposal": proposal,
                "applied": False,
                "reason": "not_enough_trades_to_experiment",
            }
        if perf.get("winRate", 0) < 30:
            return {
                "proposal": proposal,
                "applied": False,
                "reason": "win_rate_too_low_to_experiment",
            }

    result = apply_optimization(proposal)
    return {"proposal": proposal, "result": result, "applied": result.get("applied", False)}


# ── Optimization log ────────────────────────────────────────


def _load_log() -> List[Dict[str, Any]]:
    path = _optimizations_log_path()
    if not path.exists():
        return []
    try:
        with open(path) as f:
            return json.load(f)
    except Exception:
        return []


def _log_optimization(proposal: Dict[str, Any], applied: bool) -> None:
    arr = _load_log()
    arr.append(
        {
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "proposal": proposal,
            "applied": applied,
        }
    )
    arr = arr[-200:]
    path = _optimizations_log_path()
    path.parent.mkdir(parents=True, exist_ok=True)
    with open(path, "w") as f:
        json.dump(arr, f, indent=2, default=str)


def _recent_params(limit: int = 5) -> List[str]:
    arr = _load_log()
    return [e.get("proposal", {}).get("parameter") for e in arr[-limit:] if e.get("applied")]
